### PR TITLE
Enable FEATURE_BASICFREEZE and expose private QCalls

### PR DESCRIPTION
**Pull request by mjsabby**
Reflection can be used to access this experimental functionality.

My medium term goal is to get out the document for `Frozen Objects` that I've been working offline with @Maoni0 @jkotas.

### CODE REVIEW

File: src/gc/gc.h [hunk 1]
```diff
@@ -267,10 +267,6 @@ class IGCHeapInternal : public IGCHeap {
 
 protected:
```
Comment (jkotas) [inline src/gc/gc.h]: You can delete this too.

File: src/gc/gcinterface.h [hunk 2]
```diff
@@ -868,6 +868,11 @@ class IGCHeap {
     // Unregisters a frozen segment.
     virtual void UnregisterFrozenSegment(segment_handle seg) = 0;
 
+#if defined(FEATURE_BASICFREEZE) && defined(VERIFY_HEAP)
```
Comment (jkotas) [inline src/gc/gcinterface.h]: There should be no ifdefs in the GCInterface interface. Provide implementation instead.

Comment (jkotas) [inline src/gc/gcinterface.h]: This is changing GC/EE interface. We should bump the version at the top of the file to avoid LocalGC mismatches.

Comment (mjsabby) [reply src/gc/gcinterface.h]: @Maoni0 bumped it earlier today.

Comment (Maoni0) [reply src/gc/gcinterface.h]: right

File: src/gc/gc.h (hunk 1 above)
Comment (Maoni0) [reply src/gc/gc.h]: yeah you can delete both lines

`protected: `
`public:`

I think it was a merge error that they are both there.

**Comment by Maoni0** [comment]
other than the 2 minor problems @jkotas pointed above, LGTM.

File: clrdefinitions.cmake [hunk 3]
```diff
@@ -236,3 +236,5 @@ if(WIN32)
     add_definitions(-DFEATURE_DATABREAKPOINT)
   endif(CLR_CMAKE_TARGET_ARCH_AMD64 OR CLR_CMAKE_TARGET_ARCH_I386)
 endif(WIN32)
+
+add_definitions(-DFEATURE_BASICFREEZE)
```
Comment (jkotas) [inline clrdefinitions.cmake]: Could you please put it earlier in the file where other unconditional defines are?

Comment (mjsabby) [reply clrdefinitions.cmake]: Yup, put it where it says Features should be put in sorted order.

File: src/gc/gcinterface.h (hunk 2 above)
Comment (mjsabby) [reply src/gc/gcinterface.h]: Done.

File: src/gc/gc.h (hunk 1 above)
Comment (mjsabby) [reply src/gc/gc.h]: Done.

**Comment by stephentoub** [comment]
Mind sharing a sentence or two about what the feature provides? I'm intrigued by the name.

**Comment by jkotas** [comment]
https://github.com/dotnet/coreclr/issues/568#issuecomment-86609429

@mjsabby would like to experiment with this.

File: src/gc/gc.cpp [hunk 4]
```diff
@@ -3979,7 +3979,7 @@ class CObjectHeader : public Object
 #endif
         if (fSmallObjectHeapPtr)
         {
-#ifdef FEATURE_BASICFREEZE
+#if defined(FEATURE_BASICFREEZE) && defined(VERIFY_HEAP)
```
Comment (jkotas) [inline src/gc/gc.cpp]: The change in this ifdef does not look right. Why is it necessary?

**Comment by stephentoub** [comment]
> #568 (comment)

Cool, thanks.

File: src/gc/gc.cpp (hunk 4 above)
Comment (mjsabby) [reply src/gc/gc.cpp]: IsInFrozenSegment is only defined when Verify Heap is enabled.

Comment (mjsabby) [reply src/gc/gc.cpp]: But I can improve the assert condition if that's what you were getting at? I can make it this --

```cpp
#if defined(FEATURE_BASICFREEZE) && defined(VERIFY_HEAP)
            _ASSERTE(g_theGCHeap->IsInFrozenSegment(this));
#endif
            _ASSERTE(!g_theGCHeap->IsLargeObject(pMT));
```

Comment (mjsabby) [reply src/gc/gc.cpp]: Probably should, updated.

Comment (mjsabby) [reply src/gc/gc.cpp]: @Maoni0 does this look correct?

File: src/gc/gcee.cpp [hunk 5]
```diff
@@ -600,6 +600,23 @@ void GCHeap::UnregisterFrozenSegment(segment_handle seg)
 #endif // FEATURE_BASICFREEZE
 }
 
+bool GCHeap::IsInFrozenSegment(Object *object)
+{
+#if defined(VERIFY_HEAP) && defined(FEATURE_BASICFREEZE)
```
Comment (jkotas) [inline src/gc/gcee.cpp]: This should be just `#ifdef FEATURE_BASICFREEZE`.

File: src/gc/gcee.cpp [hunk 6]
```diff
@@ -600,6 +600,23 @@ void GCHeap::UnregisterFrozenSegment(segment_handle seg)
...
+    uint8_t* o = (uint8_t*)object;
+    heap_segment * hs = gc_heap::find_segment (o, FALSE);
+    //We create a frozen object for each frozen segment before the segment is inserted
+    //to segment list; during ngen, we could also create frozen objects in segments which
+    //don't belong to current GC heap.
+    //So we return true if hs is NULL. It might create a hole about detecting invalidate 
+    //object. But given all other checks present, the hole should be very small
+    return !hs || heap_segment_read_only_p (hs);
+#else // defined(VERIFY_HEAP) && defined(FEATURE_BASICFREEZE)
+    assert(!"Should not call GCHeap::IsInFrozenSegment without FEATURE_BASICFREEZE and VERIFY_HEAP defined!");
```
Comment (jkotas) [inline src/gc/gcee.cpp]: I do not think this assert is necessary.

**Comment by mjsabby** [comment]
@dotnet-bot test Windows_NT x86 Release Innerloop Build and Test

**Comment by mjsabby** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

**Comment by mjsabby** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

**Comment by mjsabby** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

**Comment by mjsabby** [comment]
Is anyone aware if the OSX10.12 x64 run is flaky? I looked at the build logs and it is a different error (some Java stack trace unable to connect to a server) than it was previously.

**Comment by jkotas** [comment]
Yes, the OSX machines are flaky.

**Pull request merged**